src/ripple/app/hook/impl/applyHook.cpp

```diff
@@ -1116,30 +1128,54 @@ hook::setHookState(
 
     if (createNew)
     {
-        ++stateCount;
+        uint32_t createdOwnerCount = computeHookStateOwnerCount(data);
+        newHookStateCount += createdOwnerCount;
 
-        if (computeHookStateOwnerCount(stateCount) > oldStateReserve)
+        if (newHookStateCount > oldHookStateCount)
         {
             // the hook used its allocated allotment of state entries for its
             // previous ownercount increment ownercount and give it another
             // allotment
 
-            ++ownerCount;
+            ownerCount += createdOwnerCount;
             XRPAmount const newReserve{view.fees().accountReserve(ownerCount)};
 
             if (STAmount((*sleAccount)[sfBalance]).xrp() < newReserve)
                 return tecINSUFFICIENT_RESERVE;
 
-            adjustOwnerCount(view, sleAccount, 1, j);
+            adjustOwnerCount(view, sleAccount, createdOwnerCount, j);
         }
 
         // update state count
-        sleAccount->setFieldU32(sfHookStateCount, stateCount);
+        sleAccount->setFieldU32(sfHookStateCount, newHookStateCount);
         view.update(sleAccount);
 
         // create an entry
         hookState = std::make_shared<SLE>(hookStateKeylet);
     }
+    else
+    {
+        // on Update
+        uint32_t oldOwnerCount =
+            computeHookStateOwnerCount((*hookState)[sfHookStateData]);
+        uint32_t newOwnerCount = computeHookStateOwnerCount(data);
+
+        uint32_t changedOwnerCount = newOwnerCount - oldOwnerCount;
+        newHookStateCount += changedOwnerCount;
+
+        if (changedOwnerCount != 0)
+        {
+            ownerCount += changedOwnerCount;
+            XRPAmount const newReserve{view.fees().accountReserve(ownerCount)};
+            if (STAmount((*sleAccount)[sfBalance]).xrp() < newReserve)
```

Comment:
TODO: Don't return `tecINSUFFICIENT_RESERVE` if the OwnerCount decreases.

Reply:
This check is not made because set_state_cache, which performs the reserve check, does not access the StateData before the change.